src/reducers/shelf/filter.ts

```diff
@@ -0,0 +1,29 @@
+import {FILTER_ADD, FILTER_REMOVE} from '../../actions/filter';
+import {Action} from '../../actions/index';
+import {DEFAULT_SHELF_UNIT_SPEC, ShelfUnitSpec} from '../../models/shelf/spec';
+import {insertItemToArray, removeItemFromArray} from '../util';
+
+export function filterReducer(shelfSpec: Readonly<ShelfUnitSpec> = DEFAULT_SHELF_UNIT_SPEC,
+                              action: Action): ShelfUnitSpec {
+  switch (action.type) {
+    case FILTER_ADD: {
+      const {filter} = action.payload;
+      const filters = insertItemToArray(shelfSpec.filters, 0, filter);
```

Comment:
`FILTER_ADD` should also take index as input (so that user can drag and drop field at a specific location). 

Or at least, we should make index = length of array (so it is appended at the end, not the beginning.)